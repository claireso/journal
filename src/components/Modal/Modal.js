import React, { useCallback, useEffect, useRef, useState } from 'react'
import PropTypes from 'prop-types'

import { IconClose } from '@components/Icons'

import * as S from './Modal.styles'

const ANIMATION_DURATION = 300

const animationsConfig = {
  open: {
    wrapper: {
      keyframes: [{ opacity: 1 }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards',
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    },
    content: {
      keyframes: [{ opacity: 1, transform: 'translate3d(0, 0, 0)' }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards',
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    }
  },
  close: {
    wrapper: {
      keyframes: [{ opacity: 0 }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards',
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    },
    content: {
      keyframes: [{ opacity: 0, transform: 'translate3d(4rem, 0, 0)' }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards',
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    }
  }
}

const Modal = ({ testId, children, onClose }) => {
  const [isOpen, setIsOpen] = useState(true)
  const wrapper = useRef()
  const content = useRef()

  const close = useCallback(() => {
    setIsOpen(false)
    setTimeout(() => {
      onClose()
    }, ANIMATION_DURATION)
  }, [onClose])

  const onClick = useCallback(
    (event) => {
      const target = event.target

      if (target.contains(content.current) && !(target.getAttribute('id') === 'modalInner')) {
        close()
      }
    },
    [close]
  )

  const onKeyDown = useCallback(
    (event) => {
      if (event && event.code === 'Escape') {
        close()
      }
    },
    [close]
  )

  useEffect(() => {
    const animations = animationsConfig[isOpen ? 'open' : 'close']

    if (wrapper.current?.animate) {
      wrapper.current.animate(animations.wrapper.keyframes, animations.wrapper.timings)
    }

    if (content.current?.animate) {
      content.current.animate(animations.content.keyframes, animations.content.timings)
    }
  }, [isOpen])

  useEffect(() => {
    document.addEventListener('keydown', onKeyDown)

    return () => {
      document.removeEventListener('keydown', onKeyDown)
    }
  }, [onKeyDown])

  return (
    <S.ModalWrapper
      ref={wrapper}
      id="modal"
      data-testid={testId}
      onClick={onClick}
      style={animationsConfig.close.wrapper.keyframes[0]}
    >
      <S.ModalInner id="modalInner" ref={content} style={animationsConfig.close.content.keyframes[0]}>
        <S.ModalCloseButton onClick={close}>
          <IconClose width="20" height="26" />
        </S.ModalCloseButton>
        {React.cloneElement(children, {
          onClose: close
        })}
      </S.ModalInner>
    </S.ModalWrapper>
  )
}

Modal.propTypes = {
  children: PropTypes.node.isRequired,
  onClose: PropTypes.func,
  testId: PropTypes.string
}

Modal.defaultProps = {
  onClose: () => {},
  testId: ''
}

export default Modal
