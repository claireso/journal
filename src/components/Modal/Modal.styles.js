import styled from 'styled-components'

import { ButtonIcon } from '../Buttons'

export const ModalWrapper = styled.div`
  background: rgba(43, 44, 44, 0.4);
  bottom: 0;
  overflow: auto;
  position: fixed;
  left: 15rem;
  right: 0;
  top: var(--toolbar-height);
  will-change: opacity;
`

export const ModalInner = styled.div`
  background: var(--white);
  box-shadow: 0 0px 11px #898c8e;
  padding: 2rem;
  max-width: 45rem;
  min-height: calc(100vh);
  margin: 0 0 0 auto;
`

export const ModalCloseButton = styled(ButtonIcon)`
  position: absolute;
  right: 1.2rem;
  top: 1.2rem;
`
