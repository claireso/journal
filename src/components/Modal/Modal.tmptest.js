// @TODO restore test

// import { render, fireEvent } from '@testing-library/react'

// import Modal from './index'

// describe('<Modal />', () => {
//   const TRANSITION_SPEED = 500

//   const renderComponent = (props = {}, options = {}) =>
//     render(
//       <Modal {...props}>
//         <div>content</div>
//       </Modal>,
//       options
//     )

//   test('should render component', () => {
//     const { container } = renderComponent({ isOpen: true, onClose: () => {} })

//     expect(container).toMatchSnapshot()
//   })

//   test('should close component by click (button close)', (done) => {
//     const props = { isOpen: true, onClose: jest.fn() }
//     const { container } = renderComponent(props)

//     fireEvent.click(container.querySelector('button'))

//     setTimeout(() => {
//       expect(props.onClose).toHaveBeenCalledTimes(1)
//       expect(container).toMatchSnapshot()
//       done()
//     }, TRANSITION_SPEED)
//   })

//   test('should close component by click (background layer)', (done) => {
//     const props = { isOpen: true, onClose: jest.fn() }
//     const { container } = renderComponent(props)

//     fireEvent.click(container.querySelector('div'))

//     setTimeout(() => {
//       expect(props.onClose).toHaveBeenCalledTimes(1)
//       expect(container).toMatchSnapshot()
//       done()
//     }, TRANSITION_SPEED)
//   })

//   test('should close component by keyboard', (done) => {
//     const props = { isOpen: true, onClose: jest.fn() }
//     const { container } = renderComponent(props)

//     fireEvent(
//       document,
//       new KeyboardEvent('keydown', {
//         key: 'Escape',
//         keyCode: 27,
//         which: 27,
//         bubbles: true
//       })
//     )

//     setTimeout(() => {
//       expect(props.onClose).toHaveBeenCalledTimes(1)
//       expect(container).toMatchSnapshot()
//       done()
//     }, TRANSITION_SPEED)
//   })

//   test('should close component by prop update', (done) => {
//     const props = { isOpen: true, onClose: jest.fn() }
//     const { container } = renderComponent(props)

//     expect(container).toMatchSnapshot()

//     renderComponent({ isOpen: false }, { container })

//     setTimeout(() => {
//       expect(container).toMatchSnapshot()
//       done()
//     }, TRANSITION_SPEED)
//   })
// })
