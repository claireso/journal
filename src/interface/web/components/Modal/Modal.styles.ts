import { styled, theme } from '@web/theme'

import { ButtonIcon } from '../Buttons'

export const Wrapper = styled('div', {
  background: 'rgba(43, 44, 44, 0.4)',
  bottom: 0,
  overflow: 'auto',
  position: 'fixed',
  left: '15rem',
  right: 0,
  top: theme.sizes.toolbarHeight.value,
  willChange: 'opacity'
})

export const Inner = styled('div', {
  background: '$white',
  boxShadow: '0 0px 11px #898c8e',
  padding: '$5',
  maxWidth: '45rem',
  minHeight: 'calc(100vh)',
  ml: 'auto'
})

export const ButtonClose = styled(ButtonIcon, {
  position: 'absolute',
  right: '1.2rem',
  top: '1.2rem'
})
