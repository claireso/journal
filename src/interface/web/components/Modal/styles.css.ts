import { cmpStyles, tokens, p, ml, responsiveStyle } from '@web/theme'

export const wrapper = cmpStyles({
  background: `color-mix(in srgb, ${tokens.colors.neutral.extradark} 50%, transparent)`,
  bottom: 0,
  overflow: 'auto',
  position: 'fixed',
  left: tokens.layout.admin.sidebar.default,
  right: 0,
  top: 0,
  willChange: 'opacity',
  zIndex: 1,
  overflowX: 'hidden',
  ...responsiveStyle({
    lg: {
      left: tokens.layout.admin.sidebar.lg
    }
  })
})

export const inner = cmpStyles({
  boxShadow: '0 0px 11px #898c8e',
  maxWidth: '50rem',
  minHeight: 'calc(100vh)',
  background: tokens.colors.white,
  overflow: 'hidden',
  ...ml('size-auto')
})

export const content = cmpStyles({
  ...p('size-4'),
  ...responsiveStyle({
    sm: {
      ...p('size-8')
    }
  })
})
