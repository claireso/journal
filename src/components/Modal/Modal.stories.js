import React, { Fragment, useState, useCallback } from 'react'
import { action } from '@storybook/addon-actions'

import Modal from './index'
import Toolbar from '../Toolbar'
import { PrimaryButton } from '../Buttons'

export default {
  title: 'Modal',
  decorators: [(storyFn) => <div style={{ margin: '-8px' }}>{storyFn()}</div>]
}

export const Basic = () => {
  const [isOpen, setOpen] = useState(false)

  const onClose = useCallback(() => {
    setOpen(false)
    action('on close')()
  }, [setOpen])

  return (
    <Fragment>
      <Toolbar>
        <PrimaryButton
          onClick={() => {
            setOpen(true)
          }}
          style={{ padding: '10px' }}
        >
          Open modal
        </PrimaryButton>
      </Toolbar>
      <Modal isOpen={isOpen} onClose={onClose}>
        <p>content</p>
      </Modal>
    </Fragment>
  )
}
