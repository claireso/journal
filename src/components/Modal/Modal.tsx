import React, { useCallback, useEffect, useRef, useState } from 'react'

import { IconClose } from '@components/Icons'

import * as S from './Modal.styles'

const ANIMATION_DURATION = 300

const animationsConfig = {
  open: {
    wrapper: {
      keyframes: [{ opacity: 1 }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards' as FillMode,
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    },
    content: {
      keyframes: [{ opacity: 1, transform: 'translate3d(0, 0, 0)' }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards' as FillMode,
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    }
  },
  close: {
    wrapper: {
      keyframes: [{ opacity: 0 }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards' as FillMode,
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    },
    content: {
      keyframes: [{ opacity: 0, transform: 'translate3d(4rem, 0, 0)' }],
      timings: {
        duration: ANIMATION_DURATION,
        fill: 'forwards' as FillMode,
        easing: 'cubic-bezier(.17,.67,.21,.97)'
      }
    }
  }
}

interface ModalProps {
  testId?: string
  children: React.ReactNode
  onClose: () => void
}

const Modal = ({ testId = '', children, onClose = () => {} }: ModalProps) => {
  const [isOpen, setIsOpen] = useState(true)
  const wrapper = useRef<HTMLDivElement>(null)
  const content = useRef<HTMLDivElement>(null)

  const close = useCallback(() => {
    setIsOpen(false)
    setTimeout(() => {
      onClose()
    }, ANIMATION_DURATION)
  }, [onClose])

  const onClick = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      const target = event.target as HTMLDivElement

      if (target.contains(content.current) && !(target.getAttribute('id') === 'modalInner')) {
        close()
      }
    },
    [close]
  )

  const onKeyDown = useCallback(
    (event: KeyboardEvent) => {
      if (event?.code === 'Escape') {
        close()
      }
    },
    [close]
  )

  useEffect(() => {
    const animations = animationsConfig[isOpen ? 'open' : 'close']

    if (wrapper.current?.animate) {
      wrapper.current.animate(animations.wrapper.keyframes, animations.wrapper.timings)
    }

    if (content.current?.animate) {
      content.current.animate(animations.content.keyframes, animations.content.timings)
    }
  }, [isOpen])

  useEffect(() => {
    document.addEventListener('keydown', onKeyDown)

    return () => {
      document.removeEventListener('keydown', onKeyDown)
    }
  }, [onKeyDown])

  return (
    <S.Wrapper
      ref={wrapper}
      id="modal"
      data-testid={testId}
      onClick={onClick}
      style={animationsConfig.close.wrapper.keyframes[0]}
    >
      <S.Inner id="modalInner" ref={content} style={animationsConfig.close.content.keyframes[0]}>
        <S.ButtonClose onClick={close}>
          <IconClose width="20" height="26" />
        </S.ButtonClose>
        {children}
      </S.Inner>
    </S.Wrapper>
  )
}

export default Modal
