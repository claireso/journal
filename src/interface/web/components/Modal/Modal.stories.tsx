import React, { Fragment, useState, useCallback } from 'react'
import type { Meta, StoryObj } from '@storybook/react-webpack5'
import { action } from 'storybook/actions'

import Modal from './Modal'
import Toolbar from '../Toolbar'
import { ButtonNeutral } from '../Buttons'

const meta: Meta<typeof Modal> = {
  title: 'Components/Modal',
  component: Modal
}

export default meta
type Story = StoryObj<typeof Modal>

export const Primary: Story = {
  render() {
    const [isOpen, setOpen] = useState(false)

    const onClose = useCallback(() => {
      setOpen(false)
      action('on close')()
    }, [setOpen])

    return (
      <Fragment>
        <Toolbar>
          <ButtonNeutral
            onClick={() => {
              setOpen(true)
            }}
            style={{ padding: '10px' }}
          >
            Open modal
          </ButtonNeutral>
        </Toolbar>
        {isOpen && (
          <Modal title="Modal title" onClose={onClose}>
            <p>content</p>
          </Modal>
        )}
      </Fragment>
    )
  }
}
