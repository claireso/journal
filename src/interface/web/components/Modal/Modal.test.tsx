import { render, screen, fireEvent, waitFor } from '@testing-library/react'

import Modal from './index'

describe('<Modal />', () => {
  const renderComponent = (props = {}, options = {}) =>
    render(
      <Modal title="Modal title" onClose={jest.fn()} {...props}>
        <div>content</div>
      </Modal>,
      options
    )

  it('should render component', () => {
    const { asFragment } = renderComponent({ isOpen: true, onClose: () => {} })

    expect(asFragment()).toMatchSnapshot()
  })

  it('should close component by click (button close)', async () => {
    const props = { isOpen: true, onClose: jest.fn() }

    renderComponent(props)

    fireEvent.click(screen.getByRole('button'))

    await waitFor(() => expect(props.onClose).toHaveBeenCalledTimes(1))
  })

  it('should close modal by click (background layer)', async () => {
    const props = { isOpen: true, onClose: jest.fn(), testId: 'modal' }

    renderComponent(props)

    fireEvent.click(screen.getByTestId(props.testId))

    await waitFor(() => expect(props.onClose).toHaveBeenCalledTimes(1))
  })

  it('should close component by keyboard', async () => {
    const props = { isOpen: true, onClose: jest.fn() }

    renderComponent(props)

    fireEvent.keyDown(document, { key: 'Escape', code: 'Escape', charCode: 27 })

    await waitFor(() => expect(props.onClose).toHaveBeenCalledTimes(1))
  })
})
