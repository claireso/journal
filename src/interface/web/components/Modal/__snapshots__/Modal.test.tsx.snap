// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render component 1`] = `
<DocumentFragment>
  <div
    class="styles_layerStyle__f6ptmm0"
    data-testid=""
    id="modal"
    style="opacity: 0;"
  >
    <div
      class="styles_layerStyle__f6ptmm1"
      id="modalInner"
      style="opacity: 0; transform: translate3d(4rem, 0, 0);"
    >
      <div
        class="styles__188a5vq2 styles_layerStyle__188a5vq0 styles_toolbar_variant_neutral__188a5vq3 styles_layerStyle__188a5vq1"
      >
        <h2
          class="styles_layerStyle__pkh2681"
        >
          Modal title
        </h2>
        <button
          class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_dark__1mdc28jj styles_layerStyle__1mdc28j4 styles_button_size_sm__1mdc28jk styles_layerStyle__1mdc28j5"
        >
          <svg
            aria-hidden="true"
            class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_sm__19o76jub styles_layerStyle__19o76ju2 styles_icon_variant_default__19o76juh styles_layerStyle__19o76ju8"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18 18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        class="styles_layerStyle__f6ptmm2"
      >
        <div>
          content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
